### ISSUE TITLE
Number recognition in table is too eager

### ISSUE THREAD

**Issue by Mc-Zen** [Mc-Zen/zero#1]
In general, `Zero` is a bit too eager recognizing the content of a cell as a number. While non-text content is (as intended) never interpreted as a number, plain text usually is. 

For example `[Text]` becomes $T\times 10^\mathrm{xt}$ because `Zero` does not actually check if the individual characters are really digits. 

However, this is also not quite unintended because numbers _can_ actually contain letters - a common case are hexadecimal numbers. 
Also, it might be useful for some applications to allow a symbolic exponent, e.g., `num[1.23ex]` which is rendered as $1.23\times 10^x$.

**Comment by aszenz** [comment]
I think it's better to disable number recognition and force users to specify exactly which columns are numeric. 

From my experience in excel which also tries to recognize numbers in columns, there are too many edge cases.

Also when using this in scripting it is entirely possible that the table output changes because for some reason the input data is now recognized as a number

**Comment by aszenz** [comment]
In addition to this, a lot of fields are actually numbers but shouldn't be formatted as such, examples are ids, invoice numbers, transaction ids etc.

**Comment by Mc-Zen** [comment]
I think, there might be a misunderstanding here: zero does not check all cells for numbers. Number conversion and alignment is already only performed for columns for which it is turned on by the user. This is achieved through the `format` parameter which takes an array of `auto` or `none` (or a dictionary), see [here](https://github.com/Mc-Zen/zero?tab=readme-ov-file#table-alignment). 

This issue is mainly concerned with things like table header cells that are not numbers in a column that _is explicitly set_ for number alignment. It's easy to avoid number recognition for individual cells by adding a space before or after (e.g., `[Angle ]` instead of `[Angle]`) but still.

**Comment by aszenz** [comment]
> This issue is mainly concerned with things like table header cells that are not numbers in a column that _is explicitly set_ for number alignment. It's easy to avoid number recognition for individual cells by adding a space before or after (e.g., `[Angle ]` instead of `[Angle]`) but still.

Should header cells even be formatted (even if they are numbers), is there no way to check that a cell is header/not in typst?

**Comment by Mc-Zen** [comment]
Good idea. Unfortunately (as far as I know), it is not possible to detect header cells. Also, I guess not everyone uses `table.header` which is mostly necessary for repeatable headers across pages.

**Comment by pedrojvaz** [comment]
I also have a case where Zero things of the Euro sign as being a digit and applies grouping. Any way to avoid that?

![image](https://github.com/user-attachments/assets/8ad15dd3-f714-4d3e-850e-6e503c05353d)
![image](https://github.com/user-attachments/assets/4c44852a-29b9-48a4-adcb-885fc5957c0c)

**Comment by Mc-Zen** [comment]
Hi @pedrojvaz ,

For me, this is actually a different (but very valid) issue. 

> In general, Zero is a bit too eager recognizing the content of a cell as a number. While non-text content is (as intended) never interpreted as a number, plain text usually is.

Here I meant the mechanism that makes the decision whether a cell *is* a number *or not*. 

On the other hand, discovering whether *part* of some content is a number (and leaving the rest as it is) is not even implemented as of now and this is what you are concerned with I believe. I would therefore suggest creating an additional, separate issue for that and invite you to do so if you would. I have collected some thoughts about how to add that feature properly and would comment that there :)

**Comment by Alex-Muirhead** [comment]
Would it be worth using multiple different patterns to try and match different supported formats, such as one for scientific notation and one for hexadecimal, and either let the user specify (with default as scientific) or do automatic detection (with a consistent format per column)?

I think using symbolic exponents etc might be niche enough that it could use a separate format, which itself could be more eager. This would allow the other formats to be more conservative.

**Comment by Mc-Zen** [comment]
Hi @Alex-Muirhead ,
This is a good suggestion. I will think a bit about it

> I think using symbolic exponents etc might be niche enough that it could use a separate format, which itself could be more eager. This would allow the other formats to be more conservative.

Now that you mention that, I think the least we can do immediately is require that numerals start with either a digit or a comma/dot. This would rule out _many_ false positives now (and reduce the cases where one needs to add a space for header cells to make them escape from the number detection). 

One thing - and maybe you can give your opinion as well:
I wonder how important hexadecimal numbers actually are to number alignment in tables. Because with the requirement that numerals need to start with `[0-9\,\.]`, those might not be recognized anymore. However, who writes hexadecimal numbers with fractional parts .... ?

**Comment by Alex-Muirhead** [comment]
I've had a play around using a bit of regex, and had a similar pattern starting with either a digit or a comma/dot, and combined with the `get.text` function from the `t4t` package it was quite robust. 

I can only say I've seen hex representations of floating points as literals in codebases (such as `0x1.4p3`), but never typeset before. We could require that any hex number starts with `0x`, to make it stricter.

**Comment by Mc-Zen** [comment]
Okay, if we require then to start with 0x, its no problem since the number is then starting with 0-9.,

**Comment by Mc-Zen** [comment]
I think with the changes from #10 it is fine now.
